### PR TITLE
 wireguard: 0.0.20190406 -> 0.0.20190531 and Change peers without tearing down the interface, handle DNS failures better

### LINKED ISSUE #62288: nixos/wireguard: should use WG_ENDPOINT_RESOLUTION_RETRIES
## Issue description

In https://github.com/NixOS/nixpkgs/pull/61971 there was an option introduced that indefinitely tries to restart the wireguard service until all
peer hostnames are resolved. Instead it should use `WG_ENDPOINT_RESOLUTION_RETRIES`
as introduced by 
https://git.zx2c4.com/WireGuard/commit?id=a1a1c1723b6423e43d71cfa52862fad9bbcd265b
instead. This feature will be available in the next snapshot release of wireguard.
cc @sjau @Ma27 @zx2c4

### PR DESCRIPTION
**Pull request by grahamc**
###### Motivation for this change

In #30459 @sjau reported a problem with starting up a system with WireGuard where peers used DNS. On boot, the system would start the WireGuard interface, and attempt to add a peer by hostname. This would fail because the system wasn't able to resolve domains yet.

In #61971 sjau fixed this bug by having the wg startup unit restart until it worked, but this caused @zx2c4 great heartache.

Relatedly, having the entire interface and all peers disappear because one peer failed to resolve is not a nice or acceptable failure mode. This is related to  another problem: changing a single peer causes the interface to go away and be restarted from scratch. This isn't necessary, either :)

 - Commit 1 changes the unit to create a systemd unit per peer
 - Commit 2 updates wireguard from  0.0.20190406 -> 0.0.20190531 which adds a flag to fix the original bug  report -- by letting resolution retry forever
 - Commit 3 takes advantage of that flag

cc @Mic92 

 - Closes #62288
 - Makes @zx2c4 happier

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/modules/services/networking/wireguard.nix [hunk 1]
```diff
@@ -265,25 +316,8 @@ let
           wg set ${name} private-key ${privKey} ${
             optionalString (values.listenPort != null) " listen-port ${toString values.listenPort}"}
 
-          ${concatMapStringsSep "\n" (peer:
-            assert (peer.presharedKeyFile == null) || (peer.presharedKey == null); # at most one of the two must be set
```
Comment (grahamc) [inline nixos/modules/services/networking/wireguard.nix]: This assert has been moved down in to the assertions, which provides a reasonable error message than "assertion failed" :P

**Comment by grahamc** [comment]
@GrahamcOfBorg build wireguard-tools
@GrahamcOfBorg test wireguard
@GrahamcOfBorg test wireguard-generated

**Review by disassembler — APPROVED** [review]

**Review by flokli — APPROVED** [review]
LGTM.

Also switched over my system to this PR.

**Comment by flokli** [comment]
We probably either want to revert ced7cfc02523e1bbc636373d24b7d16110ee587e in 19.03, or backport this PR to 19.03. I guess doing the former makes more sense.

**Comment by Mic92** [comment]
I think backporting all wireguard snapshots makes more sense. This is what I used to do in the past.

**Comment by flokli** [comment]
I wasn't referring to the backporting of wireguard snapshots, but on the change of retry behaviour introduced in the mentioned commit, which we should revert on 19.03, too.

**Comment by zx2c4** [comment]
Sorry guys, you might want to bump again to 0.0.20190601. 0531 has some issues on Intel Atom.

File: nixos/modules/services/networking/wireguard.nix [hunk 2]
```diff
@@ -229,8 +229,61 @@ let
         '';
       };
 
+  generatePeerUnit = { interfaceName, interfaceCfg, peer }:
+    let
+      keyToUnitName = replaceChars
+        [ "/" "-"    " "     "+"     "="      ]
+        [ "-" "\\x2d" "\\x20" "\\x2b" "\\x3d" ];
```
Comment (Ma27) [inline nixos/modules/services/networking/wireguard.nix]: Any reason you didn't use `utils.escapeSystemdPath` here?

**Review by Ma27 — COMMENTED** [review]
Sorry for my late response, just had a deeper look at this and IMHO this is way better now, so thanks a lot @grahamc @zx2c4!

**Comment by sjau** [comment]
Unfortunately that seems to prevent to start client interfaces properly. Having it applied, it did properly start the server interface, but several client interfaces do show up as started as well but there's no ping possible.

**Comment by sjau** [comment]
It actually seems the two commands `wg set ... peer ....` and `ip route ...` are created but in a peer unit file and that unit file isn't properly started. See:

```
[08:56] <clever> have you tried restarting the peer unit?
[08:56] <hyper_ch2> systemctl restart wireguard-wg_ons --> same result
[08:57] <clever> the peer unit, not that one
[08:57] <hyper_ch2> I have no idea how to do that
[08:57] <clever> systemctl restart wireguard-wg_ons-peer-enDIyyZaperJVAk-GxsTEHx-VlxCEZ9PM1uCvoO-Km8\x3d.service
[08:57] <clever> didnt copy the whole thing, oops
[08:58] <hyper_ch2> Failed to restart wireguard-wg_ons-peer-enDIyyZaperJVAk-GxsTEHx-VlxCEZ9PM1uCvoO-Km8x3d.service: Unit wireguard-wg_ons-peer-enDIyyZaperJVAk-GxsTEHx-VlxCEZ9PM1uCvoO-Km8x3d.service not found.
[08:58] <hyper_ch2> the backslash in there is really, really annoying
[08:58] <clever> try \\
[08:58] <clever> and i agree
[08:59] <hyper_ch2> clever: now it works
[...]
[08:59] <hyper_ch2> still, after reboot, it doesn't work anymore
[09:00] <clever> hyper_ch2: sounds like the peer units just need more wantedby flags, which would be a bug
[...]
[09:08] <hyper_ch2> clever: so, what's the solution?
[09:09] <clever> hyper_ch2: the peer units likely need to be improved, to auto-run on bootup
```

**Comment by sjau** [comment]
It seem the peer unit files are properly created - but not called by the current unit files:

```
cd /etc/systemd/sytem
ls -l *wireguard*
```
results in:
```
wireguard-wg_hb-peer-quHqlmRZb2J\x2bg2V5GgOXqERaVLYjx5vAUMgYLIgqo2U\x3d.service
wireguard-wg_hb.service
wireguard-wg_home-peer-aSzrILnaPQJ5lIMekIvGi6tsTQqgluazvhcjcwFQsFg\x3d.service
wireguard-wg_home-peer-b7m83\x2bqXz1LAGPVRrWeJ0c0AQYEKRb4VL67w1OOliVw\x3d.service
wireguard-wg_home-peer-CXgaSD\x2b4zhH2aHwCxMsAeCQqkWsz\x2bf95m10bWeBhNA8\x3d.service
wireguard-wg_home-peer-fUKlG8Rb0eHSUJYVbkHtrWuX7\x2b6XM0u\x2b5NLmygCcVBA\x3d.service
wireguard-wg_home-peer--LaoVQ2sjnk66d2RQiOn1ImMBxcMDnz7o1Virq2hLRk\x3d.service
wireguard-wg_home-peer-lcYd1Q2yX-UHW08fqI-AWJLK-wUzj5meJ0q3F8zlQiE\x3d.service
wireguard-wg_home-peer-pRuFE4CYInp7VHQre66I18brnJTKye\x2bYDXS8iQF-zDc\x3d.service
wireguard-wg_home.service
wireguard-wg_jl-peer-pRuFE4CYInp7VHQre66I18brnJTKye\x2bYDXS8iQF-zDc\x3d.service
wireguard-wg_jl.service
wireguard-wg_ons-peer-enDIyyZaperJVAk-GxsTEHx-VlxCEZ9PM1uCvoO-Km8\x3d.service
wireguard-wg_ons.service
```

Looking at `wireguard-wg_hb.service` it returns:

```
[Unit]
After=network.target network-online.target
Description=WireGuard Tunnel - wg_hb
Requires=network-online.target

[Service]
Environment="DEVICE=wg_hb"
Environment="LOCALE_ARCHIVE=/nix/store/wzrv1xk177p4zazzv26288c936375c23-glibc-locales-2.27/lib/locale/locale-archive"
Environment="PATH=/nix/store/yi7zbr0fs6622j56iwfi4446cg7k2f5j-kmod-26/bin:/nix/store/awszgbq1w4fqkg4slkl95544sg20v50v-iproute2-5.1.0/bin:/nix/store/a45vrxj0rybi9pagdx914j4wdpvpz480-wireguard-tools-0.0.20190601/bin:/nix/store/bv1lw6a2kw0mn2y3lxhi43180idx6sp9-coreutils-8.31/bin:/nix/store/s1n4vl1f3in3nacalrc3xam0vyzpsfvs-findutils-4.6.0/bin:/nix/store/7d9bi31h40hky30f5scqx7r6wn311ain-gnugrep-3.3/bin:/nix/store/qg4qbkbca7qapfzpa8p991yjf944fc3w-gnused-4.7/bin:/nix/store/v5zabkifl190b69lp92wsssn8zdp8370-systemd-239.20190219/bin:/nix/store/yi7zbr0fs6622j56iwfi4446cg7k2f5j-kmod-26/sbin:/nix/store/awszgbq1w4fqkg4slkl95544sg20v50v-iproute2-5.1.0/sbin:/nix/store/a45vrxj0rybi9pagdx914j4wdpvpz480-wireguard-tools-0.0.20190601/sbin:/nix/store/bv1lw6a2kw0mn2y3lxhi43180idx6sp9-coreutils-8.31/sbin:/nix/store/s1n4vl1f3in3nacalrc3xam0vyzpsfvs-findutils-4.6.0/sbin:/nix/store/7d9bi31h40hky30f5scqx7r6wn311ain-gnugrep-3.3/sbin:/nix/store/qg4qbkbca7qapfzpa8p991yjf944fc3w-gnused-4.7/sbin:/nix/store/v5zabkifl190b69lp92wsssn8zdp8370-systemd-239.20190219/sbin"
Environment="TZDIR=/nix/store/62wblqsm4pfs5yqsi92g82lwmspgyprh-tzdata-2019a/share/zoneinfo"



ExecStart=/nix/store/4gjzpinpamy8zs0xdbnmdafhvd7hrfpm-unit-script-wireguard-wg_hb-start 
ExecStopPost=/nix/store/2a0jyj2zkgr7gllr63mmxd44n08v5034-unit-script-wireguard-wg_hb-post-stop
RemainAfterExit=true
Type=oneshot
```


Looking at the ExecStart file `/nix/store/4gjzpinpamy8zs0xdbnmdafhvd7hrfpm-unit-script-wireguard-wg_hb-start`, we'll get this output:
```
#! /nix/store/x0mlaj4z4ciycaycfwc36l1932mwywfj-bash-4.4-p23/bin/bash -e
modprobe wireguard



ip link add dev wg_hb type wireguard

ip address add 10.100.200.2/32 dev wg_hb

wg set wg_hb private-key /nix/store/83pnx2jmkdx6y0qy28c2glzhbypcgkjf-wg-key 

ip link set up dev wg_hb
```
but as said before, two instructions for the peer (in this case the server peer) are missing. However, in `/etc/systemd/system` we also have the peer unit file for the wg_hb interface: `wireguard-wg_hb-peer-quHqlmRZb2J\x2bg2V5GgOXqERaVLYjx5vAUMgYLIgqo2U\x3d.service` with this content:
```
[Unit]
After=wireguard-wg_hb.service
Description=WireGuard Peer - wg_hb - quHqlmRZb2J+g2V5GgOXqERaVLYjx5vAUMgYLIgqo2U=
Requires=wireguard-wg_hb.service

[Service]
Environment="DEVICE=wg_hb"
Environment="LOCALE_ARCHIVE=/nix/store/wzrv1xk177p4zazzv26288c936375c23-glibc-locales-2.27/lib/locale/locale-archive"
Environment="PATH=/nix/store/awszgbq1w4fqkg4slkl95544sg20v50v-iproute2-5.1.0/bin:/nix/store/a45vrxj0rybi9pagdx914j4wdpvpz480-wireguard-tools-0.0.20190601/bin:/nix/store/bv1lw6a2kw0mn2y3lxhi43180idx6sp9-coreutils-8.31/bin:/nix/store/s1n4vl1f3in3nacalrc3xam0vyzpsfvs-findutils-4.6.0/bin:/nix/store/7d9bi31h40hky30f5scqx7r6wn311ain-gnugrep-3.3/bin:/nix/store/qg4qbkbca7qapfzpa8p991yjf944fc3w-gnused-4.7/bin:/nix/store/v5zabkifl190b69lp92wsssn8zdp8370-systemd-239.20190219/bin:/nix/store/awszgbq1w4fqkg4slkl95544sg20v50v-iproute2-5.1.0/sbin:/nix/store/a45vrxj0rybi9pagdx914j4wdpvpz480-wireguard-tools-0.0.20190601/sbin:/nix/store/bv1lw6a2kw0mn2y3lxhi43180idx6sp9-coreutils-8.31/sbin:/nix/store/s1n4vl1f3in3nacalrc3xam0vyzpsfvs-findutils-4.6.0/sbin:/nix/store/7d9bi31h40hky30f5scqx7r6wn311ain-gnugrep-3.3/sbin:/nix/store/qg4qbkbca7qapfzpa8p991yjf944fc3w-gnused-4.7/sbin:/nix/store/v5zabkifl190b69lp92wsssn8zdp8370-systemd-239.20190219/sbin"
Environment="TZDIR=/nix/store/62wblqsm4pfs5yqsi92g82lwmspgyprh-tzdata-2019a/share/zoneinfo"
Environment="WG_ENDPOINT_RESOLUTION_RETRIES=infinity"



ExecStart=/nix/store/nsg799adgzjnahhzch1a3k4r4mp9kkb1-unit-script-wireguard-wg_hb-peer-quHqlmRZb2J--x2bg2V5GgOXqERaVLYjx5vAUMgYLIgqo2U--x3d-start 
ExecStopPost=/nix/store/1pjhjl6xbf40vxkjs4zkligavmrqz0xp-unit-script-wireguard-wg_hb-peer-quHqlmRZb2J--x2bg2V5GgOXqERaVLYjx5vAUMgYLIgqo2U--x3d-post-stop
RemainAfterExit=true
Type=oneshot
```
Looking there at the ExecStart file `/nix/store/nsg799adgzjnahhzch1a3k4r4mp9kkb1-unit-script-wireguard-wg_hb-peer-quHqlmRZb2J--x2bg2V5GgOXqERaVLYjx5vAUMgYLIgqo2U--x3d-start` it contains:
```
#! /nix/store/x0mlaj4z4ciycaycfwc36l1932mwywfj-bash-4.4-p23/bin/bash -e
wg set wg_hb peer [...] endpoint [...]:51820 persistent-keepalive 25 allowed-ips 10.100.200.0/24
ip route replace 10.100.200.0/24 dev wg_hb table main
```
So that's the missing part. When running those two commands, the tunnel now works.

But no idea where the peer unit file also should be called from

**Comment by grahamc** [comment]
I can't reproduce your bug report, @sjau. Please ping me on IRC when you're able to debug further.

**Comment by grahamc** [comment]
FWIW here is the test case I wrote to try and reproduce:

```nix
import ../make-test.nix ({ lib, pkgs, ...} : let
  globalPeers = [
    {
      idx = 1;
      private = "QHWxUiMMnCNdYlUUowXOhwA9dqFj0mPW+hlrfL8+33U=";
      public = "vcLLUIzLkTuv60vV7+nMsWUDSh+R+g+YR2Jn+Kjlc1g=";
    }
    {
      idx = 2;
      private = "yEhxYq3jxZMXMBCofjGNG+GomUmrwVUHvWvuFRWC2HU=";
      public = "AMYt7rajfDZfupU8+m/uQ8C+1R++zBrI+Z8wuQylXio=";
    }
    {
      idx = 3;
      private = "mDXJoAJmFaERpaP6Wvll/FqoGTxZA0CgnL68dFs4XGM=";
      public = "UWBRfv+VsUOLIi5iD++Zurl4wx+S+A6r39flcI3jyhA=";
    }
    {
      idx = 4;
      private = "gDINget094ZZNd2AV1lFxi3wOxbrtG74oZ/7FOEY30Q=";
      public = "/aqhS+qrcRMQWdCQjnnc++hggRqU+4s26EEVlnsFcTA=";
    }
    {
      idx = 5;
      private = "SBil1s3P9Uam6zJTdm/GOjgozaoVPha1ThS49OGbO0E=";
      public = "+l2Q+6Ap7y2MavVR1SR+qMQQUSp4shwNeFjcodhr+lw=";
    }
  ];

  peersFor = self: lib.remove self globalPeers;
in {
  name = "wireguard-interesting-keys";
  meta = with pkgs.stdenv.lib.maintainers; {
    maintainers = [ ma27 grahamc ];
  };

  nodes = lib.listToAttrs (map (self: {
    name = "peer${toString self.idx}";
    value = {
      networking.useDHCP = false;
      networking.interfaces.eth1 = {
        ipv4.addresses = lib.singleton {
          address = "192.168.0.${toString self.idx}";
          prefixLength = 24;
        };
      };

      networking.wireguard.interfaces.wg0 = {
        ips = [ "10.10.10.${toString self.idx}/24" ];
        listenPort = 12345;
        privateKey = self.private;
        peers = map (peer: {
          allowedIPs = [ "10.10.10.${toString peer.idx}/32" ];
          endpoint = "192.168.0.${toString peer.idx}:12345";
          publicKey = peer.public;
          persistentKeepalive = 1;
        }) (peersFor self);
      };
    };
  }) globalPeers);

  testScript = ''
    startAll;

    # Wait for wireguard-wg0 to be up
    ${lib.concatMapStringsSep "\n" (self: "$peer${toString self.idx}->waitForUnit(\"wireguard-wg0\");") globalPeers}

    ${lib.concatMapStringsSep "\n" (self: "print $peer${toString self.idx}->succeed(\"wg\");") globalPeers}

    # Ping all the peers
    ${lib.concatStringsSep "\n"
    (lib.flatten (map
      (self: map
        (peer: "$peer${toString self.idx}->succeed(\"ping -c1 10.10.10.${toString peer.idx}\");")
        (peersFor self)
      )
      globalPeers
    ))}
  '';
})
```

My suspicion was your system wasn't working because you had keys started with a `/`, however this test case does as well.

**Comment by sjau** [comment]
On latest nixos unstable it still fails to properly start the wg peers.

**Comment by flokli** [comment]
@sjau can you open an issue with some more detail on how this fails (like log output of the failing units)?

**Pull request merged**